### PR TITLE
[pytorch][graviton][build] PT 1.12 graviton release

### PR DESCRIPTION
**Pull request by ydaiming**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by tejaschumbalkar** [comment]
Lets turn on the [graviton_mode](https://github.com/aws/deep-learning-containers/blob/master/dlc_developer_config.toml#L12) flag to `true` in order to trigger graviton builds

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+# install torch and torchvision from the PyTorch website
+RUN pip uninstall -y torch \
+ && pip install --no-cache-dir -U ${PT_INFERENCE_URL} ${PT_XLA_INFERENCE_URL}
+RUN pip install "torchvision>=0.11,<0.12"
+
+## compile and install ArmCL
+RUN git clone https://github.com/Arm-software/ComputeLibrary.git \
+ && cd ComputeLibrary \
+ && git checkout $ARMCL_BRANCH \
+ && scons Werror=0 -j8 debug=0 neon=1 opencl=0 os=linux arch=arm64-v8a build=native \
```
Comment (snadampal) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:118]: it should be
scons Werror=0 -j8 debug=0 neon=1 opencl=0 os=linux arch=armv8.2-a multi_isa=1 embed_kernels=0 os=linux build=native

File: pytorch/buildspec-graviton-1-10.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
+region: &REGION <set-$REGION-in-environment>
+framework: &FRAMEWORK pytorch
+version: &VERSION 1.10.0
```
Comment (snadampal) [inline pytorch/buildspec-graviton-1-10.yml:4]: I'm not sure why you are maintaining per version yaml files. If this is the pytorch version for the current image, then it should be 1.12.0

File: pytorch/buildspec-graviton-1-10.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+os_version: &OS_VERSION ubuntu20.04
+short_version: &SHORT_VERSION "1.10"
```
Comment (snadampal) [inline pytorch/buildspec-graviton-1-10.yml:6]: same as above, it should be 1.12.0

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && cp ${HOME_DIR}/oss_compliance/build_from_source_packages/BUILD_FROM_SOURCE_PACKAGES_LICENCES_AARCH64_IMAGES ${HOME_DIR} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl https://aws-dlc-licenses.s3.amazonaws.com/pytorch-1.10/license.txt -o /license.txt
```
Comment (snadampal) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:158]: should we create a new license file?

File: pytorch/buildspec-graviton-1-10.yml (hunk 2 above)
Comment (ydaiming) [reply pytorch/buildspec-graviton-1-10.yml:4]: This is for the legacy purpose. In case there's an update for graviton 1.10 release, this is the buildspec yml file we can copy over and use.

File: pytorch/buildspec-graviton-1-10.yml (hunk 3 above)
Comment (ydaiming) [reply pytorch/buildspec-graviton-1-10.yml:6]: This file is for legacy purpose, and not used.

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 5]
```diff
@@ -0,0 +1,162 @@
+FROM arm64v8/ubuntu:20.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG PYTHON=python3
+ARG PYTHON_VERSION=3.8.12
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_graviton/torch-1.12.0-cp38-cp38-linux_aarch64.whl
+ARG PT_XLA_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_graviton/torch_xla-1.12.0%2B82fbe57-cp38-cp38-linux_aarch64.whl
+ARG ARMCL_BRANCH=v21.05
```
Comment (snadampal) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:10]: ARMCL version should be v22.05

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 6]
```diff
@@ -0,0 +1,162 @@
+FROM arm64v8/ubuntu:20.04
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG PYTHON=python3
+ARG PYTHON_VERSION=3.8.12
+ARG PT_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_graviton/torch-1.12.0-cp38-cp38-linux_aarch64.whl
+ARG PT_XLA_INFERENCE_URL=https://aws-pytorch-unified-cicd-binaries.s3.us-west-2.amazonaws.com/r1.12.0_graviton/torch_xla-1.12.0%2B82fbe57-cp38-cp38-linux_aarch64.whl
```
Comment (snadampal) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:9]: need to pull the other torch wheels as well, eg: torch text and data

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ && pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org \
+ && ln -s /opt/conda/bin/pip /usr/local/bin/pip3 \
+ && pip install packaging==20.4 \
+    enum-compat==0.0.3 \
+    "cryptography>=3.3.2"
+
+# install torch and torchvision from the PyTorch website
+RUN pip uninstall -y torch \
+ && pip install --no-cache-dir -U ${PT_INFERENCE_URL} ${PT_XLA_INFERENCE_URL}
+RUN pip install "torchvision>=0.11,<0.12"
```
Comment (snadampal) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:112]: for PyTorch 1.12, torchvision should be 0.13.0

File: pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu [hunk 8]
```diff
@@ -0,0 +1,164 @@
+FROM arm64v8/ubuntu:20.04
```
Comment (arjkesh) [inline pytorch/inference/docker/1.12/py3/Dockerfile.graviton.e3.cpu:1]: we are now changing this name to "ec2" instead of "e3" - can you change the name of the dockerfile?

File: pytorch/buildspec-graviton-1-10.yml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+images:
+  BuildE3PytorchGravitonCPUInferencePy3DockerImage:
+    <<: *INFERENCE_REPOSITORY
+    build: &PYTORCH_CPU_INFERENCE_PY3 false
+    image_size_baseline: 10000
+    device_type: &DEVICE_TYPE cpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py38
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION , "-e3"]
```
Comment (arjkesh) [inline pytorch/buildspec-graviton-1-10.yml:36]: please change the tag to "ec2" instead of "e3"

File: pytorch/buildspec-graviton-1-10.yml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      target: torchserve-e3-entrypoint.py
+    deep_learning_container:
+      source: ../../src/deep_learning_container.py
+      target: deep_learning_container.py
+    config:
+      source: docker/build_artifacts/config.properties
+      target: config.properties
+
+images:
+  BuildE3PytorchGravitonCPUInferencePy3DockerImage:
```
Comment (arjkesh) [inline pytorch/buildspec-graviton-1-10.yml:29]: please change to "EC2" instead of e3"

File: pytorch/buildspec-graviton-1-10.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      image_type: &INFERENCE_IMAGE_TYPE inference
+      root: !join [ *FRAMEWORK, "/", *INFERENCE_IMAGE_TYPE ]
+      repository_name: &REPOSITORY_NAME !join [pr, "-", *FRAMEWORK, "-", *INFERENCE_IMAGE_TYPE, "-", graviton]
+      repository: &REPOSITORY !join [ *ACCOUNT_ID, .dkr.ecr., *REGION, .amazonaws.com/, *REPOSITORY_NAME ]
+
+context:
+  inference_context: &INFERENCE_CONTEXT
+    torchserve-e3-entrypoint:
+      source: docker/build_artifacts/torchserve-e3-entrypoint.py
+      target: torchserve-e3-entrypoint.py
```
Comment (arjkesh) [inline pytorch/buildspec-graviton-1-10.yml:20]: this has been renamed torchserve-ec2-entrypoint

**Review by snadampal — APPROVED** [review]

File: pytorch/buildspec-graviton-1-10.yml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+images:
+  BuildE3PytorchGravitonCPUInferencePy3DockerImage:
+    <<: *INFERENCE_REPOSITORY
+    build: &PYTORCH_CPU_INFERENCE_PY3 false
+    image_size_baseline: 10000
+    device_type: &DEVICE_TYPE cpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py38
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION , "-e3"]
+    docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton.e3., *DEVICE_TYPE ]
```
Comment (arjkesh) [inline pytorch/buildspec-graviton-1-10.yml:37]: dockerfile needs to be renamed to .ec2.

Comment (xncqr) [reply pytorch/buildspec-graviton-1-10.yml:37]: Noticing the yml is 1.10 and not 1.12. Also this is graviton 1.12. Would recommending renaming dockerfile to Dockerfile.graviton.cpu It will be supporting EC2 and sagemaker in the end.

Comment (RadhikaB-97) [reply pytorch/buildspec-graviton-1-10.yml]: Made couple of changes to the buildspec-graviton-1.10 file to accommodate the e3 to ec2 change!

**Review by xncqr — APPROVED** [review]
some of this will be updated when the Sagemaker part of this image is added. Approving.

**Review by arjkesh — APPROVED** [review]
